### PR TITLE
Merging repos: 'dd-shared-components-dotnet' into 'dd-trace-dotnet'.

### PR DESCRIPTION
**Pull request by macrogreg**
This is the first of several self-contained PRs targeted at moving the Shared Code (as in _shared between the Tracer and the Profiler_) from [dd-shared-components-dotnet](https://github.com/DataDog/dd-shared-components-dotnet) into [dd-trace-dotnet](https://github.com/DataDog/dd-trace-dotnet).

For this PR, I created a branch `permanent/MoveFrom_dd-shared-components-dotnet` and moved all files from the `dd-shared-components-dotnet` into there, including their complete change history. The history may be optionally squashed or preserved when merging this PR into the `master` branch. The files are all inside the `/shared` folder, so that they can be naturally merged into the planned joint directory structure.

In addition, this PR also makes the following small changes:

* Make whatever tweaks necessary so that _all_ the Shared Code projects build inside `dd-trace-dotnet`.
* Set up placeholder folders for the `/tracer` and the `/profiler` products to facilitate the directory refactoring in the subsequent PRs.
* Update ReadMe files to describe the changes.

Note that this PR keeps the directory structure within the `/shared` folder exactly as it was within the original repo. Subsequent changed will adjust this structure to joint conventions as required. This will require coordination with the profiler build configuration which makes some assumptions about the layout of the relevant folders.

This change is self-contained. It will not affect any existing Tracer components, and it will allow the overall repo merge project to proceed in several steps that can be validated independently.

This PR will be followed up with additional PR to address the following:

* Move Tracer-specific files into the `/tracer` folder.
* Ensure that changes contained to shared files that are not (yet) used by the Tracer do not require the Tracer CI validation to pass (this will be updated in the future if and when the Tracer uses more shared components)
* Joint build of components used by several products.
* Joint release package.
* Joint conventions for folder structure.
* Etc...

### CODE REVIEW

**Review by andrewlock — APPROVED** [review]
GitHub borks on 300 files, so I assume it's fine 😄

File: NOTICE [hunk 1]
```diff
@@ -1,4 +1,8 @@
-Datadog .NET Tracer for APM (dd-trace-dotnet)
-Copyright 2017 Datadog, Inc.
+Application Telemetry & Performance Monitoring Product Suite for .NET Applications
+Datadog 
+
+(dd-trace-dotnet)
+
+Copyright 2017-2021 Datadog, Inc.
```
Comment (zacharycmontoya) [inline NOTICE:6]: Can you revert this back to `Copyright 2017 Datadog, Inc`? We don't need to include the end year for it to take effect, so I'd prefer removing it

File: tracer/README.MD [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿# dd-trace-dotnet/tracer
+
+This repo ([dd-trace-dotnet](https://github.com/DataDog/dd-trace-dotnet)) contains the sources for the client-side components of the Datadog product suite for Application Telemetry Collection and Application Performance Monitoring for .NET Applications.
+
+This directory (`/tracer`) contains the source for the __Tracer for .NET__ product, which is a part of that product suite.
+
+### Copyright
+
+Copyright (c) 2017-2021 Datadog
```
Comment (zacharycmontoya) [inline tracer/README.MD:9]:
```suggestion
Copyright (c) 2017 Datadog
```

File: tracer/README.MD [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+[https://www.datadoghq.com](https://www.datadoghq.com/)
+
+### License
+
+See [license information](./LICENSE).
```
Comment (zacharycmontoya) [inline tracer/README.MD:14]: I think this needs to refer to the license at the root directory:
```suggestion
See [license information](../LICENSE).
```

File: shared/src/Shared-Src/README.MD [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿This directory and its subdirectories contains source files that are stored in this repo and are used across several projects.
+
+One of the reasons for doing it is to share the implementation across projects for APIs that should
+not be public and therefore cannot be grouped into a separate shared assembly.
+
+Formanaged code, the subdirectories here carry the name of the respective namespaces.
```
Comment (zacharycmontoya) [inline shared/src/Shared-Src/README.MD:6]:
```suggestion
For managed code, the subdirectories here carry the name of the respective namespaces.
```

File: shared/README.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+This directory (`/shared`) contains code code which is shared between between the different products of that suite (.NET Tracer, .NET Profiler, etc...).
+
+### Copyright
+
+Copyright (c) 2017-2021 Datadog
+[https://www.datadoghq.com](https://www.datadoghq.com/)
+
+### License
+
+See [license information](./LICENSE).
```
Comment (zacharycmontoya) [inline shared/README.md:14]:
```suggestion
See [license information](../LICENSE).
```

File: profiler/README.MD [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+This directory (`/profiler`) contains the source for the __Continuous Profiler for .NET__ product, which is a part of that product suite. 
+
+### Copyright
+
+Copyright (c) 2017-2021 Datadog
+[https://www.datadoghq.com](https://www.datadoghq.com/)
+
+### License
+
+See [license information](./LICENSE).
```
Comment (zacharycmontoya) [inline profiler/README.MD:14]:
```suggestion
See [license information](../LICENSE).
```

**Review by zacharycmontoya — APPROVED** [review]
I left small comments on some files, but if you want to copy over everything as-is first and apply my feedback in a later PR that would be fine. Thanks for the PR 👍🏼

File: NOTICE (hunk 1 above)
Comment (macrogreg) [reply NOTICE:6]: Addressed in https://github.com/DataDog/dd-trace-dotnet/commit/8b9642541f0ad21873f3bdd9325b6601a10bb8c9

File: tracer/README.MD (hunk 2 above)
Comment (macrogreg) [reply tracer/README.MD:9]: Addressed in https://github.com/DataDog/dd-trace-dotnet/commit/8b9642541f0ad21873f3bdd9325b6601a10bb8c9

File: tracer/README.MD (hunk 3 above)
Comment (macrogreg) [reply tracer/README.MD:14]: Addressed in https://github.com/DataDog/dd-trace-dotnet/commit/8b9642541f0ad21873f3bdd9325b6601a10bb8c9

File: shared/src/Shared-Src/README.MD (hunk 4 above)
Comment (macrogreg) [reply shared/src/Shared-Src/README.MD:6]: Addressed in https://github.com/DataDog/dd-trace-dotnet/commit/8b9642541f0ad21873f3bdd9325b6601a10bb8c9

File: shared/README.md (hunk 5 above)
Comment (macrogreg) [reply shared/README.md:14]: Addressed in https://github.com/DataDog/dd-trace-dotnet/commit/8b9642541f0ad21873f3bdd9325b6601a10bb8c9

File: profiler/README.MD (hunk 6 above)
Comment (macrogreg) [reply profiler/README.MD:14]: Addressed in https://github.com/DataDog/dd-trace-dotnet/commit/8b9642541f0ad21873f3bdd9325b6601a10bb8c9

**Pull request merged**